crud/stats/init.lua

```diff
@@ -0,0 +1,291 @@
+---- CRUD statistics module.
+-- @module crud.stats
+--
+
+local clock = require('clock')
+local checks = require('checks')
+local fiber = require('fiber')
+local fun = require('fun')
+
+local dev_checks = require('crud.common.dev_checks')
+local stash = require('crud.common.stash')
+local op_module = require('crud.stats.operation')
+local registry = require('crud.stats.local_registry')
+
+local stats = {}
+local internal = stash.get(stash.name.stats_internal)
+
+--- Check if statistics module was enabled.
+--
+-- @function is_enabled
+--
+-- @treturn boolean Returns `true` or `false`.
+--
+function stats.is_enabled()
+    return internal.is_enabled == true
+end
+
+--- Initializes statistics registry, enables callbacks and wrappers.
+--
+--  If already enabled, do nothing.
+--
+-- @function enable
+--
+-- @treturn boolean Returns `true`.
+--
+function stats.enable()
+    if stats.is_enabled() then
+        return true
+    end
+
+    internal.is_enabled = true
+    registry.init()
+
+    return true
+end
+
+--- Resets statistics registry.
+--
+--  After reset collectors are the same as right
+--  after initial `stats.enable()`.
+--
+-- @function reset
+--
+-- @treturn boolean Returns true.
+--
+function stats.reset()
+    if not stats.is_enabled() then
+        return true
+    end
+
+    registry.destroy()
+    registry.init()
+
+    return true
+end
+
+--- Destroys statistics registry and disable callbacks.
+--
+--  If already disabled, do nothing.
+--
+-- @function disable
+--
+-- @treturn boolean Returns true.
+--
+function stats.disable()
+    if not stats.is_enabled() then
+        return true
+    end
+
+    registry.destroy()
+    internal.is_enabled = false
+
+    return true
+end
+
+--- Get statistics on CRUD operations.
+--
+-- @function get
+--
+-- @string[opt] space_name
+--  If specified, returns table with statistics
+--  of operations on space, separated by operation type and
+--  execution status. If there wasn't any requests of "op" type
+--  for space, there won't be corresponding collectors.
+--  If not specified, returns table with statistics
+--  about all observed spaces.
+--
+-- @treturn table Statistics on CRUD operations.
+--  If statistics disabled, returns `{}`.
+--
+function stats.get(space_name)
+    checks('?string')
+
+    if not stats.is_enabled() then
+        return {}
+    end
+
+    return registry.get(space_name)
+end
+
+-- Hack to set __gc for a table in Lua 5.1
+-- See https://stackoverflow.com/questions/27426704/lua-5-1-workaround-for-gc-metamethod-for-tables
+-- or https://habr.com/ru/post/346892/
+local function setmt__gc(t, mt)
+    local prox = newproxy(true)
+    getmetatable(prox).__gc = function() mt.__gc(t) end
+    t[prox] = true
+    return setmetatable(t, mt)
+end
+
+local function wrap_pairs_gen(build_latency, space_name, op, gen, param, state)
```

Comment:
Here we know that `gen` is a luafun iterator and can call `gen:unwrap()` to avoid double-luafun-wrapping. One `__call` methomethod call per loop iteration instead of two. Don't know, whether it really worth to change, just my observations. Maybe it is better to keep this code a bit more general and don't assume that we'll get a luafun iterator. Feel free to resolve without actions.

Reply:
I'll resolve the comment.